### PR TITLE
add teacher student sigmoid loss

### PR DESCRIPTION
**Pull request by colourful-tree**

### CODE REVIEW

File: paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc [hunk 1]
```diff
@@ -0,0 +1,256 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
```
Comment (guru4elephant) [inline paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc]: 2018

File: paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+             "Label is a Tensor<float> with shape [N x 1]. ");
+    AddOutput("Y",
+              "(Tensor, default Tensor<float>), a 2-D tensor with shape "
+              "[N x 1]. The teacher student sigmoid loss.");
+    AddAttr<float>("soft_max_up_bound", "fp32, default 15.0").SetDefault(15.0);
+    AddAttr<float>("soft_max_lower_bound", "fp32, default -15.0")
+        .SetDefault(-15.0);
+    AddComment(R"DOC(
+TeacherStudentSigmoidLoss Operator.
+TeacherStudentSigmoidLoss Operator.
```
Comment (typhoonzero) [inline paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc]: Here's a duplicate line.

File: paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+             "This input is a probability computed by the previous operator, "
+             "which is almost always the result of a softmax operator.");
+    AddInput("Label",
+             "(Tensor), the ground truth which is a 2-D tensor. "
+             "Label is a Tensor<float> with shape [N x 1]. ");
+    AddOutput("Y",
+              "(Tensor, default Tensor<float>), a 2-D tensor with shape "
+              "[N x 1]. The teacher student sigmoid loss.");
+    AddAttr<float>("soft_max_up_bound", "fp32, default 15.0").SetDefault(15.0);
+    AddAttr<float>("soft_max_lower_bound", "fp32, default -15.0")
```
Comment (typhoonzero) [inline paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc]: These two attributes' description is too short

File: paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+        when z' is not exist, clk = 0 : label = -2;
+        when z' is not exist, clk = 1 : label = -1;
+        when z' is exist    , clk = 0 : label = 0 + z';
+        when z' is exist    , clk = 1 : label = 1 + z';
+
+)DOC");
+  }
+};
+
+// template <typename DeviceContext, typename T>
```
Comment (typhoonzero) [inline paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc]: No comment out code please

File: paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+              "(Tensor, default Tensor<float>), a 2-D tensor with shape "
+              "[N x 1]. The teacher student sigmoid loss.");
+    AddAttr<float>("soft_max_up_bound", "fp32, default 15.0").SetDefault(15.0);
+    AddAttr<float>("soft_max_lower_bound", "fp32, default -15.0")
+        .SetDefault(-15.0);
+    AddComment(R"DOC(
+TeacherStudentSigmoidLoss Operator.
+TeacherStudentSigmoidLoss Operator.
+
+It's similarity to SigmoidCrossEntropyWithLogits Operator. The difference is that
```
Comment (typhoonzero) [inline paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc:129]: This operator is similar to ...

File: paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+we add another label(z') to original.
+        loss = max(x, 0) - x * z + log(1 + exp(-abs(x))) + max(x, 0) - x * z' + log(1 + exp(-abs(x)))
+        z is click or not
+        z' is value q of feed_fine
```
Comment (typhoonzero) [inline paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc]: value q, feed fine is not described.

File: paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+)DOC");
+  }
+};
+
+// template <typename DeviceContext, typename T>
+template <typename T>
+class TeacherStudentSigmoidLossOpKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    PADDLE_ENFORCE(platform::is_cpu_place(context.GetPlace()),
```
Comment (typhoonzero) [inline paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc]: when kernel is registered cpu only, no check here is needed.

Comment (typhoonzero) [reply paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc]: While this kernel have much common with `sigmoid_cross_entropy_with_logits`, can define a math functor and use it in these two ops.

File: paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+        when z' is not exist, clk = 1 : label = -1;
+        when z' is exist    , clk = 0 : label = 0 + z';
+        when z' is exist    , clk = 1 : label = 1 + z';
+
+)DOC");
+  }
+};
+
+// template <typename DeviceContext, typename T>
+template <typename T>
```
Comment (typhoonzero) [inline paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc]: template class should put in .h files

File: python/paddle/fluid/tests/unittests/test_teacher_student_sigmoid_loss_op.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+            elif label < 0.0:
+                outs.append(max(x, 0.0) - x + log(1.0 + exp(-abs(x))))
+            elif label < 1.0:
+                outs.append(max(x, 0.0) + log(1.0 + exp(-abs(x))) + \
+                            max(x, 0.0) - x * label + log(1.0 + exp(-abs(x))))
+                #print "33 python x:", x, "python label:", label, "term1:", max(x, 0.0) + log(1.0 + exp(-abs(x))), "term2:", max(x, 0.0) - x * label + log(1.0 + exp(-abs(x)))
+            else:
+                outs.append(max(x, 0.0) - x + log(1.0 + exp(-abs(x))) + \
+                            max(x, 0.0) - x * (label - 1.0) + log(1.0 + exp(-abs(x))))
+                #print "44 python x:", x, "python label:", label, "term1:", max(x, 0.0) - x + log(1.0 + exp(-abs(x))), "term2:", max(x, 0.0) - x * (label - 1.0) + log(1.0 + exp(-abs(x)))
```
Comment (typhoonzero) [inline python/paddle/fluid/tests/unittests/test_teacher_student_sigmoid_loss_op.py]: no commented out code please.

File: python/paddle/fluid/tests/unittests/test_teacher_student_sigmoid_loss_op.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        self.outputs = {'Y': np.array(outs)}
+
+    def test_check_output(self):
+        """
+            ut
```
Comment (typhoonzero) [inline python/paddle/fluid/tests/unittests/test_teacher_student_sigmoid_loss_op.py]: this docstring is not needed.

File: python/paddle/fluid/layers/nn.py [hunk 11]
```diff
@@ -9185,6 +9186,47 @@ def log_loss(input, label, epsilon=1e-4, name=None):
...
+                                 soft_max_up_bound=15.0,
+                                 soft_max_lower_bound=-15.0):
+    """
+    **Teacher Student Log Loss Layer**
+
+    This layer accepts input predictions and target label and returns the
+    teacher_student loss.
+
+    .. math::
+        loss = max(x, 0) - x * z + log(1 + exp(-abs(x))) + max(x, 0) - x * z' + log(1 + exp(-abs(x)))
```
Comment (reyoung) [inline python/paddle/fluid/layers/nn.py:9200]: It seems that all operators of this loss are exits in CPP operators. why not just combine these operators as a layer?

BTW, what is `z'` means?

Comment (colourful-tree) [reply python/paddle/fluid/layers/nn.py:9200]: z' is teacher value. the value of z' and label:
label = {-2, -1, [0, 2]}
when z' is not exist, clk = 0 : label = -2;
when z' is not exist, clk = 1 : label = -1;
when z' is exist    , clk = 0 : label = 0 + z';
when z' is exist    , clk = 1 : label = 1 + z';

File: python/paddle/fluid/tests/unittests/test_teacher_student_sigmoid_loss_op.py (hunk 10 above)
Comment (colourful-tree) [reply python/paddle/fluid/tests/unittests/test_teacher_student_sigmoid_loss_op.py]: ok, done. when push the code to icode it's nedded before

File: python/paddle/fluid/tests/unittests/test_teacher_student_sigmoid_loss_op.py (hunk 9 above)
Comment (colourful-tree) [reply python/paddle/fluid/tests/unittests/test_teacher_student_sigmoid_loss_op.py]: ok, done.

File: paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc (hunk 8 above)
Comment (colourful-tree) [reply paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc]: ok, done

File: paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc (hunk 7 above)
Comment (colourful-tree) [reply paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc]: ok, has remove this check

File: paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc (hunk 6 above)
Comment (colourful-tree) [reply paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc]: ok, modify z' is value of teacher, call teacher is more common

File: paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc (hunk 2 above)
Comment (colourful-tree) [reply paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc]: ok, done

File: paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc (hunk 3 above)
Comment (colourful-tree) [reply paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc]: ok. has modify

File: paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc (hunk 4 above)
Comment (colourful-tree) [reply paddle/fluid/operators/teacher_student_sigmoid_loss_op.cc]: ok, done

**Comment by typhoonzero** [comment]
@colourful-tree I quite agree with @reyoung 's comments, can this loss layer using current operators, so that no extra need to add operator.

**Review by panyx0718 — APPROVED** [review]
LG API
